commons/src/test/java/io/aiven/kafka/tieredstorage/commons/storage/s3/S3OutputStreamTest.java

```diff
@@ -0,0 +1,293 @@
+import com.amazonaws.services.s3.model.CompleteMultipartUploadResult;
+import com.amazonaws.services.s3.model.InitiateMultipartUploadRequest;
+import com.amazonaws.services.s3.model.InitiateMultipartUploadResult;
+import com.amazonaws.services.s3.model.PartETag;
+import com.amazonaws.services.s3.model.UploadPartRequest;
+import com.amazonaws.services.s3.model.UploadPartResult;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.ArgumentMatchers;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.mockito.Mockito.doNothing;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class S3OutputStreamTest {
+
+    static final String BUCKET_NAME = "some_bucket";
+
+    static final String FILE_KEY = "some_key";
+
+    static final String UPLOAD_ID = "some_upload_id";
+
+    @Mock
+    AmazonS3 mockedS3;
+
+    @Captor
+    ArgumentCaptor<CompleteMultipartUploadRequest> completeMultipartUploadRequestCaptor;
+    @Captor
+    ArgumentCaptor<AbortMultipartUploadRequest> abortMultipartUploadRequestCaptor;
+    @Captor
+    ArgumentCaptor<UploadPartRequest> uploadPartRequestCaptor;
+
+    final Random random = new Random();
+
+    @Test
+    void noRequestsForEmptyBytes() throws Exception {
+        try (final var out = new S3OutputStream(BUCKET_NAME, FILE_KEY, 10, mockedS3)) {
+            out.write(new byte[] {});
+        }
+
+        verify(mockedS3, never()).initiateMultipartUpload(ArgumentMatchers.any(InitiateMultipartUploadRequest.class));
+        verify(mockedS3, never()).uploadPart(ArgumentMatchers.any(UploadPartRequest.class));
+        verify(mockedS3, never()).completeMultipartUpload(ArgumentMatchers.any(CompleteMultipartUploadRequest.class));
+        verify(mockedS3, never()).abortMultipartUpload(ArgumentMatchers.any(AbortMultipartUploadRequest.class));
+    }
+
+    @Test
+    void sendAbortForAnyExceptionWhileWriting() {
+        when(mockedS3.initiateMultipartUpload(ArgumentMatchers.any(InitiateMultipartUploadRequest.class)))
+            .thenReturn(newInitiateMultipartUploadResult());
+        doNothing().when(mockedS3).abortMultipartUpload(abortMultipartUploadRequestCaptor.capture());
+
+        when(mockedS3.uploadPart(ArgumentMatchers.any(UploadPartRequest.class)))
+            .thenThrow(RuntimeException.class);
+
+        assertThatThrownBy(() -> {
+            try (final var out = new S3OutputStream(BUCKET_NAME, FILE_KEY, 100, mockedS3)) {
+                out.write(new byte[] {1, 2, 3});
+            }
+        }).isInstanceOf(IOException.class);
```

Comment:
A bit more checking:
```suggestion
        final RuntimeException testException = new RuntimeException("test");
        when(mockedS3.uploadPart(ArgumentMatchers.any(UploadPartRequest.class)))
            .thenThrow(testException);

        assertThatThrownBy(() -> {
            try (final var out = new S3OutputStream(BUCKET_NAME, FILE_KEY, 100, mockedS3)) {
                out.write(new byte[] {1, 2, 3});
            }
        }).isInstanceOf(IOException.class)
            . hasCause(testException);
```